Cloud and Enterprise Docs (#5575)
treeverse/lakeFS: docs/reference/sso.md

Reviewer: Given the length of it, I might consider splitting this doc into two. 

1. Single Sign On (SSO) - Cloud
2. Single Sign On (SSO) - Enterprise

Just a thought

Author: did, and placed it under the new Enterprise and Cloud folders :)

### Okta
{: .d-inline-block }
Cloud
{: .label .label-green }

Steps:
1. Login to your Okta account
2. Select **Applications > Applications**, then **Create App Integration**.
3. Select Create New App and enter the following:
    1. For **Sign-in method**, choose OIDC.
    2. Under **Application type**, choose **Web app**.
    3. Select Next.
4. Under General Settings:
    1. **App integration name**, enter a name for your application. (i.e lakeFS Cloud)
5. In the **Sign-in redirect URIs** field, enter https://lakefs-cloud.us.auth0.com/login (United States) or https://lakefs-cloud.eu.auth0.com/login (Europe).
6. Under **Sign-in redirect URIs**, click **Add URI**, enter https://lakefs-cloud.us.auth0.com/login/callback (United States) or https://lakefs-cloud.eu.auth0.com/login/callback (Europe).
7. Under **Assignments**, choose the wanted **Controlled access**. (i.e Allow everyone in your organization to access)
8. Uncheck **Enable immediate access with Federation Broker Mode**.

> 9. Select **Save**.

Author: Mostly because we have application specific settings in our steps (e.g the redirect URI, the "Enable immediate access with Federation Broker Mode" checkbox).
These steps focus on exactly what's needed for lakeFS.cloud compared to [Okta's guide](https://help.okta.com/en-us/Content/Topics/Apps/Apps_App_Integration_Wizard_OIDC.htm) which covers much more steps and configuration options.
I think that maybe adding a reference to that guide for whoever who wants to use the full guide, might help, I'm not sure about replacing it altogether.
WDYT?

## Enterprise

lakeFS Enterprise provides a sidecar microservice that handles the authentication called Fluffy.
Once you've onboarded to [lakeFS Enterprise](../enterprise.md) and have access to the authentication sidecar, you can configure one of the supported authentication methods below.
lakeFS Enterprise leverage Helm's capabilities to run the sidecar with it's configuration, on top of the specific-authentication method, please review the [Helm configuration](#helm).

### AD FS (using SAML)
{: .d-inline-block }
Enterprise
{: .label .label-purple }

In order for fluffy to work, some values must be configured, update the relevant values section in this file and comment it out.
1. Replace `fluffy.saml_rsa_public_cert` and `fluffy.saml_rsa_private_key` with real certificate values
2. Replace `fluffyConfig.auth.saml.idp_metadata_url` to the metadata URL of the provider <adfs-auth.company.com>
3. Replace `fluffyConfig.auth.saml.external_user_id_claim_name` with the claim name representing user id name in ADFS

> 

Reviewer 2: Maybe specify in general that in all examples `https://lakefs.company.com` is a single ingress URL to lakeFS?

Author: I've added that comment for the integrations that it wasn't explicitly mentioned

parent: Reference
nav_order: 65
has_children: false
---

# Single Sign On (SSO)
{: .d-inline-block }
Cloud
{: .label .label-green }

Enterprise
{: .label .label-purple }
{: .no_toc }

{: .note}
> SSO is only available for [lakeFS Cloud](../cloud.md) and [lakeFS Enterprise](../enterprise.md).
>
> Using the Open Source? Read more on [authentication](authentication.html). 

{% include toc.html %}

## Supported Protocols and Third Parties

lakeFS Cloud and lakeFS Enterprise support the commonly used protocols for authentication, such as OpenID Connect and SAML.

There are specific authentication utilities that we've already verified that are working, if you're using one that is not listed below but is based on OpenID Connect or SAML, it's probably a small effort for us to make sure it'll work, please [contact us](support@treeverse.io) for more information.

## Cloud

lakeFS Cloud leverage Auth0's capabilities for authentication, therefore, we're supporting Auth0's identity providers, such as: Active Directory/LDAP, ADFS, Azure Active Directory Native, Google Workspace, OpenID Connect, Okta, PingFederate, SAML and Azure Active Directory.

### Okta

> {: .d-inline-block }
> Cloud
> {: .label .label-green }
> 

Reviewer: I think these are redundant, given that it's under the "Cloud" heading

```suggestion

```

Steps:
1. Login to your Okta account
2. Select **Applications > Applications**, then **Create App Integration**.
3. Select Create New App and enter the following:
    1. For **Sign-in method**, choose OIDC.
    2. Under **Application type**, choose **Web app**.
    3. Select Next.
4. Under General Settings:
    1. **App integration name**, enter a name for your application. (i.e lakeFS Cloud)
5. In the **Sign-in redirect URIs** field, enter https://lakefs-cloud.us.auth0.com/login (United States) or https://lakefs-cloud.eu.auth0.com/login (Europe).
6. Under **Sign-in redirect URIs**, click **Add URI**, enter https://lakefs-cloud.us.auth0.com/login/callback (United States) or https://lakefs-cloud.eu.auth0.com/login/callback (Europe).
7. Under **Assignments**, choose the wanted **Controlled access**. (i.e Allow everyone in your organization to access)
8. Uncheck **Enable immediate access with Federation Broker Mode**.
9. Select **Save**.

Once you finish registering your application with Okta, save the **Client ID**, **Client Secret** and your **Okta Domain**, send this to Treeverse's team to finish the integration.

### Active Directory Federation Services (AD FS)

> {: .d-inline-block }
> Cloud
> {: .label .label-green }

Reviewer: ```suggestion

```

You add and modify redirect URIs for your registered applications by configuring their platform settings.

Enter https://lakefs-cloud.us.auth0.com/login/callback or https://lakefs-cloud.eu.auth0.com/login/callback (depends on your organization data location) as your redirect URI.

Settings for each application type, including redirect URIs, are configured in Platform configurations in the Azure portal. Some platforms, like Web and Single-page applications, require you to manually specify a redirect URI. For other platforms, like mobile and desktop, you can select from redirect URIs generated for you when you configure their other settings.

Steps:
1. In the Azure portal, in App registrations, select your application.
2. Under Manage, select Authentication.
3. Under Platform configurations, select Add a platform.
4. Under Configure platforms, select the web option.
5. Select Configure to complete the platform configuration.

Once you finish registering lakeFS Cloud with Azure AD, save the **Application (Client) ID**, **Application Secret Value** and send this to Treeverse's team to finish the integration.

## Enterprise

> lakeFS Enterprise provides a secondary service, running side-by-side with lakeFS which handles the authentication, this service is called Fluffy.

Reviewer: Is exposing this internal name deliberate (like [graveler](https://docs.lakefs.io/understand/glossary.html#graveler)) and if so are we going to write it up somewhere?

Author: due to the fact this is the only place Fluffy is being used right now, I thought mentioning it here would be sufficient.
do you think we should have a centralized place with our codenames and what service they represent?

Reviewer: It's up to you folk if you want to expose it, or just abstract it out in the docs to something like "lakeFS authorisation proxy" (or whatever it is). There are pros and cons to keeping it

Pros
- keeps the software personable and not boring
- less mental translation for devs when writing docs and talking to users

Cons
- depending on its use / surface area, moves the mental workload to users who have to figure out wtf `fluffy` and `graveler` are (but in practice not, because they're relatively low-level things that aren't day-to-day user accessed)

I'd say keep it, but at a minimum add an entry for it to the glossary.

> Once you've onboarded to [lakeFS Enterprise](../enterprise.md) and have access to the authentication service (Fluffy), you can configure one of the supported authentication methods below.

Reviewer: ```suggestion
Once you've setup [lakeFS Enterprise](../enterprise.md) and have access to the authentication service (Fluffy), you can configure one of the supported authentication methods below.
```

> lakeFS Enterprise leverage Helm's capabilities to run the sidecar with it's configuration, on top of the specific-authentication method, please review the [Helm configuration](#helm).

Reviewer: This sentence doesn't make sense. I don't understand what it's saying so I can't rewrite it. Is it two sentences that have run together? 

```suggestion
lakeFS Enterprise uses Helm's capabilities to run the sidecar with its configuration, on top of the specific-authentication method, please review the [Helm configuration](#helm).
```

Author: rephrased it, please let me know if it's still doesn't make sense

### AD FS (using SAML)
{: .d-inline-block }
Enterprise
{: .label .label-purple }

> > **Note**: AD FS integration is using certificates to sign & encrypt requests going out from Fluffy and decrypt incoming requests from AD FS server.

Reviewer: ```suggestion
{: .note}
> AD FS integration uses certificates to sign & encrypt requests going out from Fluffy and decrypt incoming requests from AD FS server.
```

> In order for fluffy to work, some values must be configured, update the relevant values section in this file and comment it out.

Reviewer: I don't understand what `this file` refers to, and why would you update them and then comment them out? 

```suggestion
In order for fluffy to work, the following values must be configured. Update the relevant values section in this file and comment it out.
```

Author: fixed, thanks

1. Replace `fluffy.saml_rsa_public_cert` and `fluffy.saml_rsa_private_key` with real certificate values
2. Replace `fluffyConfig.auth.saml.idp_metadata_url` to the metadata URL of the provider <adfs-auth.company.com>
3. Replace `fluffyConfig.auth.saml.external_user_id_claim_name` with the claim name representing user id name in ADFS

> 4. Replace `lakefs.company.com` with the lakeFS server endpoint.

Author: in the configuration detailed below

> 4. Replace `lakefs.company.com` with the lakeFS server endpoint.

Reviewer: that needs to be made clearer

If you'd like to generate the certificates using OpenSSL, you can take a look at the following example:
```sh
openssl req -x509 -newkey rsa:2048 -keyout myservice.key -out myservice.cert -days 365 -nodes -subj "/CN=lakefs.company.com" -
```

> lakeFS Enterprise Configuration:

Reviewer: Should this be "lakeFS Server configuration" ?

Author: it's the helm values file, updated

```yaml

> lakefsConfig: |

Reviewer: is this a line of the yaml? is the `|` deliberate?

Author: yes, this is part of the helm configuration, this is how we instruct multi line configuration

  # Important: make sure to include the rest of your lakeFS Configuration here!

  auth:
    cookie_auth_verification:
      auth_source: saml
      friendly_name_claim_name: displayName
      external_user_id_claim_name: samName
      default_initial_groups:
        - "Developers"
    logout_redirect_url: "https://lakefs.company.com/logout-saml"
    encrypt:
      secret_key: shared-secrey-key
    ui_config:
      login_url: "https://lakefs.company.com/sso/login-saml"
      logout_url: "https://lakefs.company.com/sso/logout-saml"
      login_cookie_names:
        - internal_auth_session

>         - saml_auth_session

Reviewer: is there a better way to articulate how to deal with this bleeding chunk of configuration? for example, "Replace the existing `auth` section of your lakeFS server config with the following" ?

      rbac: external
```

Fluffy Configuration:

> ```yaml

Reviewer: is this a file? what's the path?

Author: updated with a comment that it refers to the values.yaml file

fluffyConfig: &fluffyConfig |
  logging:
    format: "json"
    level: "INFO"
    audit_log_level: "INFO"
    output: "="
  installation:
    fixed_id: fluffy-authenticator
  auth:  
    # better set from secret FLUFFY_AUTH_ENCRYPT_SECRET_KEY must be equal to what is used in lakeFS for auth_encrypt_secret_key
    encrypt:
      secret_key: shared-secrey-key    
    logout_redirect_url: https://lakefs.company.com
    post_login_redirect_url: https://lakefs.company.com
    saml:
      enabled: true 
      sp_root_url: https://lakefs.company.com
      sp_x509_key_path: '/etc/saml_certs/rsa_saml_private.cert'
      sp_x509_cert_path: '/etc/saml_certs/rsa_saml_public.pem'
      sp_sign_request: true
      sp_signature_method: "http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"
      idp_metadata_url: "https://adfs-auth.company.com/federationmetadata/2007-06/federationmetadata.xml"
      # idp_authn_name_id_format: "urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified"
      external_user_id_claim_name: samName
      # idp_metadata_file_path: 

>       # idp_skip_verify_tls_cert: true

Reviewer: are these supposed to be commented out?

Author: they're optional for people who'd like to set these configuration, they're not must and it really depends on how the third party (e.g ldap, oidc, etc) is configured

```

> ### OpenID Connect

Reviewer: I wonder if these sections would be better presented as tabs (like we use in the current quickstart)
![CleanShot 2023-03-29 at 21 53 55](https://user-images.githubusercontent.com/3671582/228664853-e06a9b31-b2f3-4202-9a09-6a6367443fc3.png)

The page gets awfully long to scroll down otherwise.

```yaml
{% raw %}###############################################
###############################################
############ COMMON CONFIGURATION #############
###############################################
###############################################

fluffy:
  serviceAccountName: default
  name: &fluffyName lakefs-fluffy
  logLevel: INFO
  replicaCount: &fluffyReplicaCount 1
  resources: &resources
    limits:
      memory: 256Mi
    requests:
      memory: 128Mi
  ingress:
    enabled: &enableIngress true 
    host: &ingressHost lakefs.company.com
    ingressClassName: &ingressClassName ''
    annotations: &ingressAnnotations {}
  image: 
    repository: 'treeverse/fluffy'
    tag: '0.1.1'
    secretName: 'fluffy-dockerhub'
    secretToken: &fluffyImageSecretToken '<docker token>'
  saml_rsa_public_cert: |
    -----BEGIN CERTIFICATE-----
    -----END CERTIFICATE-----
  saml_rsa_private_key: |
    -----BEGIN PRIVATE KEY-----
    -----END PRIVATE KEY-----

###############################################
###############################################
###### ADVANCED DON'T EDIT VALUES BELOW #######
###############################################

> ###############################################

Reviewer: why are we putting this in the docs if all the user has to do is copy and paste it and not change it? wouldn't it be better to ship a config file and then tell them where to find it and what to change?

Author: they can edit the part below if they like, it's advanced configuration, we are planning on releasing a dedicated chart for the SSO service that will reduce the need of these configuration snippets
